tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,362 @@
+
+## Do Virtual Machines Dream of Electric Sheep?
+
+Our problem here is that a virtual machine (VM) in a typical configuration
+barely sees any physical hardware. Users do not interact with VMs in server
+scenarios using a directly-connected keyboard and mouse. VMs make use of a
+virtualized network interface and virtualized disks. Even the CPU features can
+be intercepted and virtualized. Virtual environments are entropy-starved.
+
+This is a known problem and [various][randomness-exposed]
+[analyses][not-so-random] of the weakness of random outputs in virtual
+environments have been published. The problem is especially severe right after
+boot. The gradual trickle of unpredictability from hardware events slowly moves
+the pseudo-random stream into an unpredictable state, and at the very start, it
+tends to be fairly predictable. Typically, operating systems store some of their
+PRNG output on shutdown and use it to quickly reseed their PRNG on the next
+boot, in order to reuse whatever entropy was contained in its state.
+Unfortunately, it is common to boot several machines from the same system image,
+or from a pristine image lacking a seed, making random outputs in a virtual
+machine vulnerable to prediction close to the startup phase.
+
+To help solve these problems, we employ several sources of entropy in MirageOS
+unikernels. The case of a Unix executable is simple, as we reuse the system's
+own RNG, as exposed via `/dev/urandom`, as the source of our entropy. This is
+because the kernel is in a much better position to enter an unpredictable state
+than a process running underneath. The case of Xen unikernel is harder. Here, we
+group the entropy sources into those that originate within the unikernel itself,
+and those that originate from outside of it.
+
+In the external case, we again rely on the kernel interacting with the hardware,
+but this time it's the dom0 kernel. We have a background service,
+[Xentropyd][xentropyd], which runs in dom0, reads the RNG and serves its output
+to other domains through the Xen Console. The problem is that in many scenarios,
+like hosting on popular cloud providers, we cannot expect extra cooperation from
+dom0. A further problem is that although most of the code is there, we haven't
+fully fleshed out this design and it remains disabled in MirageOS 2.5.0.
+
+So we need to be able to achieve unpredictability relying purely on what is
+available inside a unikernel. A unikernel has no direct exposure to the
+hardware, but it is of course interacting with the outside world. To tap into
+this ambiental entropy, we have to continuously sample all inter-event timings
+in its event loop. This process is analogous to what happens in a full-blown OS
+kernel, except our events lack the extra hardware context, and our timers are
+potentially less granular (for example, on ARM). This makes our
+interaction-based events somewhat more predictable, or in other words, they have
+a little less entropy.
+
```

Comment:
Well the general game with entropy harvesting is: the more, the merrier. It would not hurt.

But there is a difference between timing and that. Timing of events is generally not fully known to anyone, even an attacker causing them to happen. Clocks are slightly out of sync, and event scheduling depends on Xen scheduling, unikernel entering the poll, and random CPU jitter. So any single reading has at least a little bit of entropy, and we known how to effectively mix that.

Depending on circumstances, packet content can be completely known to an attacker.

OTOH in high-traffic scenarios, I can imagine the total description of all incoming traffic to be at least a little bit unknown, so I guess you could actually use it as a source.